src/fairseq2/recipes/hf/datasets.py

```diff
@@ -0,0 +1,188 @@
+        **extra: Any,
+    ) -> DataPipelineReader[HFBatch]:
+        """
+        Convert HF Dataset into a fairseq2 PipelineReader.
+        The HF-specific processing logic is assumed to be
+        done before the conversion.
+
+        **Batching**: The batching can be done externally or using
+        bucketing strategies in fairseq2. If both ``bucket_size``
+        and ``max_num_elements`` are None, then we assume the
+        ``dataset`` is already batched and do nothing. Otherwise:
+
+        - If ``bucket_size`` is not None, generate batches of fixed size
+        - If ``max_num_elements`` is not None, generate batches up to
+        `max_num_elements` size. In this case, ``max_seq_len`` must be
+        given.
+
+        In addition, the data
+        should be prepared to have a column that contains
+        sequence length for each example.
+
+        This function should not be used for big datatset.
+
+        :param gang:
+            The gang over which to shard the dataset.
+        :param max_num_elements:
+            The maximum number of elements in each batch.
+        :param max_seq_len:
+            The maximum sequence length of each example.
+            Examples longer than this value will be cropped.
+        :param min_seq_len:
+            The minimum sequence length of each example. Examples
+            shorter than this value will be dropped.
+        :param seq_len_col:
+            The column that stores length of source sequence. Must
+            be provided if the data is to be batched by length
+        :param num_accumulate:
+            The number of batches to accumulate in each iteration.
+            Typically used with gradient accumulation during
+            training.
+        :param num_prefetch:
+            The number of batches to prefetch in background.
+        :param seed:
+            The seed to initialize the random number generators
+            used internally.
+        :param extras:
+            The extra parameters specific to the dataset
+            implementation.
+        """
+
+        # Resolve the batching strategies
+        def resolve_batching_strategy() -> int:
+            if bucket_size is None and max_num_elements is None:
+                return 0  # NO_BATCH
+            elif bucket_size is not None and max_num_elements is not None:
+                raise ValueError("Can only batch by size or by length, not both")
+            elif bucket_size is not None:
+                return 1  # BATCH_BY_SIZE
+            else:
+                assert (
+                    max_seq_len
+                ), "Max_seq_len must be specified in batch-by-length mode"
+                return 2  # BATCH_BY_LEN
+
+        batch_mode = resolve_batching_strategy()
+        if batch_mode == 0:
+            data = dataset.data.to_batches()  # list of RecordBatch
+        else:
+            data = dataset.to_list()
+        builder = read_sequence(data)
+
+        seed += 1
```

Comment:
We don't really have any RNG operation up to this point. This line is not necessary.